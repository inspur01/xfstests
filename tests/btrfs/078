#! /bin/bash
# FSQA Test No. 078
#
# Regression test for a btrfs issue where creation of readonly snapshots caused
# the filesystem to get into an inconsistent state.
#
# This regression was introduced in the 3.17 kernel and fixed by reverting the
# following linux kernel commit:
#
#     Btrfs: race free update of commit root for ro snapshots
#     9c3b306e1c9e6be4be09e99a8fe2227d1005effc
#
#-----------------------------------------------------------------------
#
# Copyright (C) 2014 SUSE Linux Products GmbH. All Rights Reserved.
# Author: Filipe Manana <fdmanana@suse.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it would be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write the Free Software Foundation,
# Inc.,  51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#-----------------------------------------------------------------------
#

seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"

tmp=`mktemp -d`
status=1	# failure is the default!
trap "_cleanup; exit \$status" 0 1 2 3 15

_cleanup()
{
	rm -fr $tmp
}

# get standard environment, filters and checks
. ./common/rc
. ./common/filter

# real QA test starts here
_supported_fs btrfs
_supported_os Linux
_require_scratch

rm -f $seqres.full

workout()
{
	ops=$1
	procs=$2
	num_snapshots=$3

	_scratch_mkfs >> $seqres.full 2>&1
	_scratch_mount

	snapshot_cmd="$BTRFS_UTIL_PROG subvolume snapshot -r $SCRATCH_MNT"
	snapshot_cmd="$snapshot_cmd $SCRATCH_MNT/snap_\`date +'%H_%M_%S_%N'\`"
	run_check $FSSTRESS_PROG -p $procs \
	    -x "$snapshot_cmd" -X $num_snapshots -d $SCRATCH_MNT -n $ops
}

ops=8000
procs=4
snapshots=100
workout $ops $procs $snapshots

# The fstests framework runs a file system check against the scratch device
# automatically when a test case finishes (if the test calls _require_scratch).
# That filesystem check (btrfsck, btrfs.fsck) failed reporting several fs
# inconsistencies. Therefore there's no need to call _check_scratch_fs here.

echo "Silence is golden"
status=0
exit
