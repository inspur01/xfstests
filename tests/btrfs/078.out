QA output created by 078
Silence is golden
